src/blk/kernel/KernelDevice.cc

```diff
@@ -1107,18 +1107,21 @@ int KernelDevice::aio_write(
   return 0;
 }
 
-int KernelDevice::_discard(uint64_t offset, uint64_t len)
+int KernelDevice::discard(uint64_t offset, uint64_t len)
 {
   int r = 0;
   if (cct->_conf->objectstore_blackhole) {
     lderr(cct) << __func__ << " objectstore_blackhole=true, throwing out IO"
 	       << dendl;
     return 0;
   }
-  dout(10) << __func__
-	   << " 0x" << std::hex << offset << "~" << len << std::dec
-	   << dendl;
-  r = BlkDev{fd_directs[WRITE_LIFE_NOT_SET]}.discard((int64_t)offset, (int64_t)len);
+  if (support_discard) {
```

Comment:
Now I think this isn't ideal location for this check - allocator's foreach() call will still make tons of no-op calls of this function.
So may be it's worth to have an additional global check before starting trimming for a specific device

Reply:
Yes, I have already realized.